bundles/org.openhab.binding.luftdateninfo/src/main/java/org/openhab/binding/luftdateninfo/internal/handler/HTTPHandler.java

```diff
@@ -0,0 +1,159 @@
+/**
+ * Copyright (c) 2010-2020 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.luftdateninfo.internal.handler;
+
+import java.time.LocalDateTime;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.jetty.client.HttpClient;
+import org.eclipse.jetty.client.api.ContentResponse;
+import org.openhab.binding.luftdateninfo.internal.dto.SensorData;
+import org.openhab.binding.luftdateninfo.internal.dto.SensorDataValue;
+import org.openhab.binding.luftdateninfo.internal.utils.DateTimeUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+
+/**
+ * The {@link HTTPHandler} is responsible for HTTP requests and JSON handling
+ *
+ * @author Bernd Weymann - Initial contribution
+ */
+@NonNullByDefault
+public class HTTPHandler {
+    private final Logger logger = LoggerFactory.getLogger(HTTPHandler.class);
+
+    private static final Gson GSON = new Gson();
+    private static final HTTPHandler HTTP_HANDLER = new HTTPHandler();
+
+    public static final String P1 = "P1";
+    public static final String P2 = "P2";
+
+    public static final String TEMPERATURE = "temperature";
+    public static final String HUMIDITY = "humidity";
+    public static final String PRESSURE = "pressure";
+    public static final String PRESSURE_SEALEVEL = "pressure_at_sealevel";
+
+    public static final String NOISE_EQ = "noise_LAeq";
+    public static final String NOISE_MIN = "noise_LA_min";
+    public static final String NOISE_MAX = "noise_LA_max";
+
+    private static String sensorUrl = "http://data.sensor.community/airrohr/v1/sensor/";
+    private static @Nullable HttpClient commonHttpClient;
+
+    public static void init(HttpClient httpClient) {
+        commonHttpClient = httpClient;
+    }
+
+    public static HTTPHandler getHandler() {
+        return HTTP_HANDLER;
+    }
+
+    public @Nullable String getResponse(int sensorId)
+            throws InterruptedException, TimeoutException, ExecutionException {
+        HttpClient localClient = commonHttpClient;
+        if (localClient == null) {
+            logger.warn("HTTP Client not initialized");
+            return null;
+        } else {
+            String url = sensorUrl + sensorId + "/";
+            ContentResponse contentResponse = localClient.newRequest(url).timeout(180, TimeUnit.SECONDS).send();
```

Comment:
180 sec. seem a bit long for the timeout. The framework's safe caller will kick in after 15 sec.

10 sec. would be a better value.

Reply:
Ok, let me leave then the 15 seconds timeout.

Reply:
That might lead to a race condition, as the safe caller timeout is 15sec, too. If you want to use higher timeouts, you could use Jetty's async API.

Reply:
Ok, let me get this straight because I'm implementing a binding for BMW ConnectedDrive which is based on REST API too and I don't want to fall in the same trap again.

I switched back to my previous timeout of 30 seconds and measured the duration.
```
2020-08-28 14:26:58.322 [WARN ] [o.internal.handler.BaseSensorHandler] - Sensor 39745 call duration 30006 ms
2020-08-28 14:26:58.385 [INFO ] [o.internal.handler.BaseSensorHandler] - Total timeout 30000 ms elapsed / 2020-08-28 14:26:58
```
I didn't observe a "safe caller timeout" after 15 seconds. I also don't know in which race condition can happen in/after this synchronous call. 

I've also taken a look into the Jetty Asynchronous API => https://www.eclipse.org/jetty/documentation/current/http-client-api.html#http-client-async . In general I see the use cases "large content download" and "parallel requests" which is in my opinion not fitting to this sensor binding.

So my honest question why it's suitable to apply a timeout like 1 Minute with the above observations?
I really like to have your honest opinion.

Reply:
The safe caller only kicks in if the code is invoked by the framework e.g. via `handleCommand()`. I just saw that your code will be invoked only within scheduler threads. These have no time limit. But that doesn't mean that you should block them longer as needed, as the threads in the scheduler pool are shared amongst all ThingHandlers of all bindings.

I think a timeout of more than 10 sec. is a bit unreasonable for most use cases. There was a use case with an embedded device, which was really slow in responding. A timeout of 30 sec. was sufficient. So, I'm wondering what use case you have in mind when specifying an even higher timeout.

Jetty's async API is not only useful for "large content downloads, for parallel processing of requests/responses", but also if "efficient thread and resource utilization is a key factor". The last points are quite important, as openHAB often runs on small hardware like raspberries.

IMHO event-driven program execution should be preferred over blocking, where possible.

Reply:
I fully agree! I played around with async API and I really like the unblocking character. I'll perform changes accordingly.

Reply:
I don't want to push you, but the merge window for 2.5.9 comes closer and closer. If we miss that, the binding must be ported to OH3 manually.

Reply:
No problem. Sorry, I'm too busy with my next contribution. But this was a very good hint because it's also related to get data from REST API.

I applied the changes and tested them. As said I like the unblocking character and also makes the error handling more comprehensive.